src/main/java/servlets/UploadReceiptServlet.java

```diff
@@ -53,8 +54,33 @@
   private static final String DEV_SERVER_BASE_URL = "http://0.0.0.0:80";
   // Matches JPEG image filenames.
   private static final Pattern validFilename = Pattern.compile("([^\\s]+(\\.(?i)(jpe?g))$)");
-  // Logs to System.err by default.
-  private static final Logger logger = Logger.getLogger(UploadReceiptServlet.class.getName());
+
+  private static Logger logger;
+  private static BlobstoreService blobstoreService;
+  private static BlobInfoFactory blobInfoFactory;
+  private static DatastoreService datastore;
+  private static Clock clock;
+
+  public UploadReceiptServlet() {}
+
+  public UploadReceiptServlet(Logger logger, BlobstoreService blobstoreService,
+      BlobInfoFactory blobInfoFactory, DatastoreService datastore, Clock clock) {
+    this.logger = logger;
+    this.blobstoreService = blobstoreService;
+    this.blobInfoFactory = blobInfoFactory;
+    this.datastore = datastore;
+    this.clock = clock;
+  }
+
+  @Override
+  public void init() {
+    // Logs to System.err by default.
+    logger = Logger.getLogger(UploadReceiptServlet.class.getName());
+    blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    blobInfoFactory = new BlobInfoFactory();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    clock = Clock.systemDefaultZone();
+  }
 
```

Comment:
nit: since this is a refactoring, for future reference, I'd put this in a separate pull request.